pkg/expr/nodes.go

```diff
@@ -191,69 +192,100 @@ func (s *Service) buildDSNode(dp *simple.DirectedGraph, rn *rawNode, req *Reques
+			req.Queries = append(req.Queries, backend.DataQuery{
 				RefID:         dn.refID,
 				MaxDataPoints: dn.maxDP,
 				Interval:      time.Duration(int64(time.Millisecond) * dn.intervalMS),
 				JSON:          dn.query,
 				TimeRange:     dn.timeRange.AbsoluteTime(now),
 				QueryType:     dn.queryType,
-			},
-		},
-		Headers: dn.request.Headers,
-	}
-
-	responseType := "unknown"
-	respStatus := "success"
-	defer func() {
-		if e != nil {
-			responseType = "error"
-			respStatus = "failure"
-			span.AddEvents([]string{"error", "message"},
-				[]tracing.EventValue{
-					{Str: fmt.Sprintf("%v", err)},
-					{Str: "failed to query data source"},
-				})
+			})
 		}
-		logger.Debug("Data source queried", "responseType", responseType)
-		useDataplane := strings.HasPrefix(responseType, "dataplane-")
-		s.metrics.dsRequests.WithLabelValues(respStatus, fmt.Sprintf("%t", useDataplane), dn.datasource.Type).Inc()
-	}()
 
-	resp, err := s.dataService.QueryData(ctx, req)
-	if err != nil {
-		return mathexp.Results{}, MakeQueryError(dn.refID, dn.datasource.UID, err)
-	}
+		responseType := "unknown"
+		respStatus := "success"
+		defer func() {
+			if e != nil {
+				responseType = "error"
+				respStatus = "failure"
+				span.AddEvents([]string{"error", "message"},
+					[]tracing.EventValue{
+						{Str: fmt.Sprintf("%v", err)},
+						{Str: "failed to query data source"},
+					})
+			}
+			logger.Debug("Data source queried", "responseType", responseType)
+			useDataplane := strings.HasPrefix(responseType, "dataplane-")
+			s.metrics.dsRequests.WithLabelValues(respStatus, fmt.Sprintf("%t", useDataplane), firstNode.datasource.Type).Inc()
+		}()
 
-	dataFrames, err := getResponseFrame(resp, dn.refID)
-	if err != nil {
-		return mathexp.Results{}, MakeQueryError(dn.refID, dn.datasource.UID, err)
-	}
+		resp, err := s.dataService.QueryData(ctx, req)
+		if err != nil {
+			return MakeQueryError(firstNode.refID, firstNode.datasource.UID, err)
+		}
 
-	var result mathexp.Results
-	responseType, result, err = convertDataFramesToResults(ctx, dataFrames, dn.datasource.Type, s, logger)
-	if err != nil {
-		err = MakeConversionError(dn.refID, err)
+		for _, dn := range nodeGroup {
+			dataFrames, err := getResponseFrame(resp, dn.refID)
+			if err != nil {
+				return MakeQueryError(dn.refID, dn.datasource.UID, err)
+			}
+
+			var result mathexp.Results
+			responseType, result, err = convertDataFramesToResults(ctx, dataFrames, dn.datasource.Type, s, logger)
+			if err != nil {
+				return MakeConversionError(dn.refID, err)
+			}
+			vars[dn.refID] = result
+		}
 	}
-	return result, err
+	return nil
+}
+
+// Execute runs the node and adds the results to vars. If the node requires
+// other nodes they must have already been executed and their results must
+// already by in vars.
+func (dn *DSNode) Execute(ctx context.Context, now time.Time, _ mathexp.Vars, s *Service) (r mathexp.Results, e error) {
+	panic("Execute called on DSNode and should not be")
```

Comment:
It looks like a violation of the strategy pattern currently implemented. What do you think about extending DSNode to support multiple RefIDs? The Execute can return map[RefID]Results.

Reply:
An alternative could be changing the design to use a 'controller' pattern and keep node implementation as a container of query and data, which is what is implemented in this PR but needs to be expanded to other nodes.

Reply:
The part that took me a bit to wrap my head around is that this doesn't change the dependency graph.

So for example, A->B, C->D, where both A and C are different queries to the same datasource - even though we group those DS queries into the same call for execution and run them first (since we know they are always root nodes), the actual dependency structure (The required nodes for B and D) remains unchanged. (Note, even though we group them, it is still possible for query A to fail and query C to succeed).

Reply:
I think it does probably mean that the interfaces needed to be changed such that I can remove Execute on Node, and then do a typeSwitch. (Maybe executable node). But with this and the PR for changing how errors are returned, I'm thinking it is maybe just to pile on the debt for now and tackle that (eventually) after.

Reply:
Ok. I do not want to postpone the change for the design's sake. Let's create an issue to tackle the tech debt and move forward with this PR.